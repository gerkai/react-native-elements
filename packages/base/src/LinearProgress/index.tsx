import { LinearProgress, LinearProgressProps } from './LinearProgress';

/**
 * ### Linear Progress
 *
 * Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates. They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
 *
 * Usage ```<LinearProgress variant={LinearProgress.INDETERMINATE}  />```
 *
 * [API Documentation](https://reactnativeelements.com/docs/linearProgress)
 */
const Default = Object.assign(LinearProgress, {
  INDETERMINATE: 'indeterminate' as LinearProgressProps['variant'],
  DETERMINATE: 'determinate' as LinearProgressProps['variant'],
});

export { Default as LinearProgress };
export type { LinearProgressProps };
