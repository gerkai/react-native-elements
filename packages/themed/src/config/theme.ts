import { Colors } from './colors';
import { ComponentTheme } from './theme.component';

export type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export type ThemeMode = 'light' | 'dark';

export interface ThemeSpacing {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  xl: number;
}

export interface Theme {
  mode: ThemeMode;
  spacing: ThemeSpacing;
}

export interface FullTheme extends ComponentTheme, Theme {
  colors: Colors;
}
