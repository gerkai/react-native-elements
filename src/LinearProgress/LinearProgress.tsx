import React from 'react';
import {
  View,
  StyleProp,
  ViewStyle,
  Animated,
  Platform,
  ViewProps,
} from 'react-native';
import Color from 'color';
import { RneFunctionComponent } from '../helpers';

export type LinearProgressProps = ViewProps & {
  /** The value of the progress indicator for the determinate variant. Value between 0 and 1. */
  value?: number;

  /** Type of button. */
  variant?: 'determinate' | 'indeterminate';

  /** Color for linear progress. */
  color?: 'primary' | 'secondary' | string;

  /** Track color for linear progress. */
  trackColor?: string;

  /** Add additional styling for linear progress component. */
  style?: StyleProp<ViewStyle>;
};

/** Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates.
 * They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
 * Also receives all [View](https://reactnative.dev/docs/view#props) props */
export const LinearProgress: RneFunctionComponent<LinearProgressProps> = ({
  value = 0,
  variant = 'indeterminate',
  color = 'secondary',
  style,
  theme,
  trackColor,
  ...props
}) => {
  const [width, setWidth] = React.useState<number>(0);

  const { current: animation } = React.useRef<Animated.Value>(
    new Animated.Value(0)
  );

  const intermediate = React.useRef<Animated.CompositeAnimation>();

  const startAnimation = React.useCallback(() => {
    if (variant === 'indeterminate') {
      intermediate.current = Animated.timing(animation, {
        duration: 2000,
        toValue: 1,
        useNativeDriver: true,
        isInteraction: false,
      });
      animation.setValue(0);

      Animated.loop(intermediate.current).start();
    } else {
      Animated.timing(animation, {
        duration: 1000,
        toValue: value || 0,
        useNativeDriver: Platform.OS !== 'web',
        isInteraction: false,
      }).start();
    }
  }, [animation, variant, value]);

  const tintColor =
    color === 'secondary' || color === 'primary'
      ? theme?.colors?.[color]
      : Color(color).rgb().string() || theme?.colors?.secondary;

  const trackTintColor =
    trackColor || Color(tintColor).alpha(0.4).rgb().string();

  React.useEffect(() => {
    startAnimation();
  }, [startAnimation, value]);

  return (
    <View
      accessible
      accessibilityRole="progressbar"
      accessibilityValue={{
        now: value,
        min: 0,
        max: 1,
      }}
      {...props}
      onLayout={(e) => {
        setWidth(e.nativeEvent.layout.width);
      }}
      style={[
        {
          height: 4,
          overflow: 'hidden',
          width: '100%',
          backgroundColor: trackTintColor,
        },
        style,
      ]}
    >
      <Animated.View
        style={{
          transform: [
            {
              translateX: animation.interpolate(
                variant === 'indeterminate'
                  ? {
                      inputRange: [0, 1],
                      outputRange: [-width, 0.5 * width],
                    }
                  : {
                      inputRange: [0, 1],
                      outputRange: [-0.5 * width, 0],
                    }
              ),
            },
            {
              scaleX: animation.interpolate(
                variant === 'indeterminate'
                  ? {
                      inputRange: [0, 0.5, 1],
                      outputRange: [0.0001, 1, 0.001],
                    }
                  : {
                      inputRange: [0, 1],
                      outputRange: [0.0001, 1],
                    }
              ),
            },
          ],
          backgroundColor: tintColor as string,
          flex: 1,
        }}
      />
    </View>
  );
};

LinearProgress.displayName = 'LinearProgress';
