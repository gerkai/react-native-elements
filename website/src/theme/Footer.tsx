import Link from '@docusaurus/Link';
import React from 'react';
import useThemeContext from '@theme/hooks/useThemeContext';
import Sponsor from '../components/Sponsor';

const communities = [
  {
    name: 'Twitter',
    url: 'https://twitter.com/nativebase',
  },
  {
    name: 'StackOverflow',
    url: 'https://stackoverflow.com/questions/tagged/native-base',
  },
  {
    name: 'Discord',
    url: 'https://discord.com/invite/TSgCw2UPmb',
  },
  {
    name: 'VS Code Extension',
    url: 'https://marketplace.visualstudio.com/items?itemName=NativeBase.nativebase-v3-vscode-extension&ssr=false#review-details',
  },
];
const docs = [
  {
    name: 'Introduction',
    url: 'https://docs.nativebase.io/',
  },
  {
    name: 'Examples',
    url: 'https://docs.nativebase.io/todo-example',
  },
];
const more = [
  {
    name: 'GitHub',
    url: 'https://github.com/GeekyAnts/NativeBase',
  },
  {
    name: 'Contribution Guidelines',
    url: 'https://github.com/GeekyAnts/NativeBase/blob/master/CONTRIBUTING.md',
  },
  {
    name: 'Privacy Policy',
    url: 'https://nativebase.io/privacy-policy',
  },
  {
    name: 'Cookie Policy',
    url: 'https://nativebase.io/cookie-policy',
  },
];
export default function Footer() {
  const { isDarkTheme } = useThemeContext();
  const darkFooter = isDarkTheme ? 'footer-dark' : '';
  return (
    <footer className={'footer ' + darkFooter}>
      <div className="container container-fluid">
        <div className="row footer__links">
          <div className="col footer__col">
            <div className="footer__title">Docs</div>
            <ul className="footer__items">
              <li className="footer__item">
                <a className="footer__link-item" href="/docs">
                  Getting Started
                </a>
              </li>
              <li className="footer__item">
                <a className="footer__link-item" href="/docs/overview">
                  Components
                </a>
              </li>
            </ul>
          </div>
          <div className="col footer__col">
            <div className="footer__title">Contribution</div>
            <ul className="footer__items">
              <li className="footer__item">
                <a
                  className="footer__link-item"
                  href="/docs/repo/contributing#setup"
                >
                  Setup Guide
                </a>
              </li>
              <li className="footer__item">
                <a className="footer__link-item" href="/docs/repo/contributing">
                  Contribution Guide
                </a>
              </li>
            </ul>
          </div>
          <div className="col footer__col">
            <div className="footer__title">Community</div>
            <ul className="footer__items">
              <li className="footer__item">
                <a
                  href="https://react-native-elements-slack.herokuapp.com/"
                  target="_blank"
                  rel="noopener noreferrer"
                  className="footer__link-item"
                >
                  Chat with us on Slack
                </a>
              </li>
              <li className="footer__item">
                <a
                  href="https://github.com/react-native-elements/react-native-elements/issues/"
                  target="_blank"
                  rel="noopener noreferrer"
                  className="footer__link-item"
                >
                  Submit a bug or feature
                </a>
              </li>
              <li className="footer__item">
                <a href="/#sponsor" className="footer__link-item">
                  Our Sponsors
                </a>
              </li>
            </ul>
          </div>
          <div className="col footer__col">
            <div className="footer__title">More</div>
            <ul className="footer__items">
              <li className="footer__item">
                <a
                  href="https://github.com/react-native-elements/react-native-elements"
                  target="_blank"
                  rel="noopener noreferrer"
                  className="footer__link-item"
                >
                  GitHub
                </a>
              </li>
              <li className="footer__item">
                <a
                  href="https://stackoverflow.com/questions/tagged/react-native-elements"
                  target="_blank"
                  rel="noopener noreferrer"
                  className="footer__link-item"
                >
                  StackOverflow
                </a>
              </li>
              <li className="footer__item">
                <a
                  href="https://marketplace.visualstudio.com/items?itemName=rne.snippets"
                  target="_blank"
                  rel="noopener noreferrer"
                  className="footer__link-item"
                >
                  VS Code Extension
                </a>
              </li>
            </ul>
          </div>
        </div>
        <div className="footer__bottom text--center">
          <div className="margin-bottom--sm">
            <Link
              to="https://opencollective.com/react-native-elements"
              className="button button--secondary button--outline button--md  margin-horiz--sm margin-vert--sm "
            >
              Open Collective
              <svg
                style={{ margin: '0 5px -3px 8px ' }}
                width={18}
                fill="#7FADF2"
                role="img"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>Open Collective</title>
                <path d="M12 0C5.373 0 0 5.373 0 12s5.373 12 12 12c2.54 0 4.894-.79 6.834-2.135l-3.107-3.109a7.715 7.715 0 1 1 0-13.512l3.107-3.109A11.943 11.943 0 0 0 12 0zm9.865 5.166l-3.109 3.107A7.67 7.67 0 0 1 19.715 12a7.682 7.682 0 0 1-.959 3.727l3.109 3.107A11.943 11.943 0 0 0 24 12c0-2.54-.79-4.894-2.135-6.834z" />
              </svg>
            </Link>
            <Link
              to="https://github.com/sponsors/react-native-elements"
              className="button button--secondary button--outline button--md margin-horiz--sm margin-vert--sm "
            >
              GitHub Sponsor
              <svg
                style={{ margin: '0 5px -3px 8px ' }}
                width={16}
                fill="#EA4AAA"
                role="img"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>GitHub Sponsors</title>
                <path d="M17.625 1.499c-2.32 0-4.354 1.203-5.625 3.03-1.271-1.827-3.305-3.03-5.625-3.03C3.129 1.499 0 4.253 0 8.249c0 4.275 3.068 7.847 5.828 10.227a33.14 33.14 0 0 0 5.616 3.876l.028.017.008.003-.001.003c.163.085.342.126.521.125.179.001.358-.041.521-.125l-.001-.003.008-.003.028-.017a33.14 33.14 0 0 0 5.616-3.876C20.932 16.096 24 12.524 24 8.249c0-3.996-3.129-6.75-6.375-6.75zm-.919 15.275a30.766 30.766 0 0 1-4.703 3.316l-.004-.002-.004.002a30.955 30.955 0 0 1-4.703-3.316c-2.677-2.307-5.047-5.298-5.047-8.523 0-2.754 2.121-4.5 4.125-4.5 2.06 0 3.914 1.479 4.544 3.684.143.495.596.797 1.086.796.49.001.943-.302 1.085-.796.63-2.205 2.484-3.684 4.544-3.684 2.004 0 4.125 1.746 4.125 4.5 0 3.225-2.37 6.216-5.048 8.523z" />
              </svg>
            </Link>
          </div>
          <div className="margin-bottom--sm">
            <img
              src="/img/logo.png"
              alt=""
              className="themedImage_node_modules-@docusaurus-theme-classic-lib-next-theme-ThemedImage-styles-module themedImage--dark_node_modules-@docusaurus-theme-classic-lib-next-theme-ThemedImage-styles-module footer__logo"
            />
          </div>
        </div>
      </div>
      <div className="text--center"></div>
    </footer>
  );
}
