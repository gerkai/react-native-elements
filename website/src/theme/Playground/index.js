/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
import clsx from 'clsx';
import Translate from '@docusaurus/Translate';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import BrowserOnly from '@docusaurus/BrowserOnly';
import { usePrismTheme } from '@docusaurus/theme-common';
import styles from './styles.module.css';
import useIsBrowser from '@docusaurus/useIsBrowser';
import { MdCode, MdOpenInNew } from 'react-icons/md';
//  import type {Props} from '@theme/Playground';
//  import type {ThemeConfig} from '@docusaurus/theme-live-codeblock';

function Header({ children }) {
  return <div className={clsx(styles.playgroundHeader)}>{children}</div>;
}

function LivePreviewLoader() {
  // Is it worth improving/translating?
  return <div>Loading...</div>;
}

function ResultWithHeader({ setOpen }) {
  return (
    <>
      {/* <Header>
        <Translate
          id="theme.Playground.result"
          description="The result label of the live codeblocks"
        >
          Result
        </Translate>
      </Header> */}
      {/* https://github.com/facebook/docusaurus/issues/5747 */}
      <div className={styles.playgroundPreview}>
        <BrowserOnly fallback={<LivePreviewLoader />}>
          {() => (
            <>
              <LivePreview />
              <LiveError />
              <MdCode color="white" />
              <button onClick={setOpen}>
                <MdCode color="white" />
              </button>
            </>
          )}
        </BrowserOnly>
      </div>
    </>
  );
}

function ThemedLiveEditor() {
  const isBrowser = useIsBrowser();
  return (
    <LiveEditor
      // We force remount the editor on hydration,
      // otherwise dark prism theme is not applied
      key={String(isBrowser)}
      className={styles.playgroundEditor}
    />
  );
}

function EditorWithHeader() {
  const [open, setOpen] = React.useState(false);

  const handleToggle = () => {
    setOpen((isOpen) => !isOpen);
  };
  return (
    <>
      <div className={styles.playgroundPreview}>
        <BrowserOnly fallback={<LivePreviewLoader />}>
          {() => (
            <>
              <LivePreview />
              <LiveError />
              <div className={clsx(styles.toggleIcon)}>
                <MdCode onClick={handleToggle} />{' '}
                {/* <MdOpenInNew onClick={openInSnack} /> */}
              </div>
            </>
          )}
        </BrowserOnly>
      </div>
      {open && (
        <>
          {' '}
          <Header>
            <Translate
              id="theme.Playground.liveEditor"
              description="The live editor label of the live codeblocks"
            >
              Live Editor
            </Translate>
          </Header>
          <ThemedLiveEditor />
        </>
      )}
    </>
  );
}

export default function Playground({ children, transformCode, ...props }) {
  const {
    siteConfig: { themeConfig },
  } = useDocusaurusContext();
  const {
    liveCodeBlock: { playgroundPosition },
  } = themeConfig;
  const prismTheme = usePrismTheme();

  return (
    <div className={styles.playgroundContainer}>
      {/* @ts-expect-error: type incompatibility with refs */}
      <LiveProvider
        code={children.replace(/\n$/, '')}
        transformCode={transformCode || ((code) => code)}
        theme={prismTheme}
        {...props}
      >
        {playgroundPosition === 'top' ? (
          <>
            <EditorWithHeader />
          </>
        ) : (
          <>
            <EditorWithHeader />
          </>
        )}
      </LiveProvider>
    </div>
  );
}
