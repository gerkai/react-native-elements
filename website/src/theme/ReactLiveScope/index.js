/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import * as RNE from 'react-native-elements';
import * as RNEUI from '@rneui/base';
import * as RNEUI_Layout from '@rneui/layout';
import LinearGradient from 'react-native-linear-gradient';

const StateLessSwitch = () => {
  const [state, setState] = React.useState(false);
  return <RNEUI.Switch value={state} onValueChange={setState} />;
};

// Add react-live imports you need here
const ReactLiveScope = {
  React,
  LinearGradient,
  ...RNE,
  ...RNEUI,
  ...RNEUI_Layout,
  Switch: StateLessSwitch,
  ...React,
};

export default ReactLiveScope;
