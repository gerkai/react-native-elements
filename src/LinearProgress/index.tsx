import { withTheme } from '../config';
import { LinearProgress, LinearProgressProps } from './LinearProgress';

export { LinearProgress };
export type { LinearProgressProps };
/**
 * ### Linear Progress
 *
 * Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates. They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
 *
 * Usage ```<LinearProgress variant={LinearProgress.INDETERMINATE}  />```
 *
 * [API Documentation](https://reactnativeelements.com/docs/linearProgress)
 */
export default Object.assign(withTheme(LinearProgress, 'LinearProgress'), {
  INDETERMINATE: 'indeterminate' as LinearProgressProps['variant'],
  DETERMINATE: 'determinate' as LinearProgressProps['variant'],
});
