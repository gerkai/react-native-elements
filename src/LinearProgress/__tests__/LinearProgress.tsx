import React from 'react';
import LinearProgress from '../index';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { render } from '@testing-library/react-native';
import theme from '../../config/theme';

describe('LinearProgress Component', () => {
  beforeEach(() => {
    let useEffect = jest.spyOn(React, 'useEffect');
    useEffect.mockImplementation((f) => f());
    const app = shallow(<LinearProgress theme={theme} />);
    expect(app.length).toBe(1);
  });

  it('should render without issues', () => {
    const app = shallow(<LinearProgress theme={theme} />);
    expect(app.length).toBe(1);
    expect(toJson(app)).toMatchSnapshot();
  });

  it('should render indeterminant without issues', () => {
    const component = shallow(
      <LinearProgress theme={theme} color="red" trackColor="blue" />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render determinant without issues', () => {
    const component = shallow(
      <LinearProgress
        variant="determinate"
        value={0.4}
        theme={theme}
        color="red"
      />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should contain the required accessibility properties', () => {
    const component = render(<LinearProgress value={0.4} />);
    const progressBar = component.getByA11yRole('progressbar');
    expect(progressBar.props).toMatchObject({
      accessibilityValue: {
        now: 0.4,
        min: 0,
        max: 1,
      },
    });
  });
});
