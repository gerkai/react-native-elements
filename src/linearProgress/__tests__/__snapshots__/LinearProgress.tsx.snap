// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearProgress Component should render determinant without issues 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityValue={
    Object {
      "max": 1,
      "min": 0,
      "now": 0.4,
    }
  }
  accessible={true}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "rgba(255, 0, 0, 0.4)",
        "height": 4,
        "overflow": "hidden",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Object {
        "backgroundColor": "rgb(255, 0, 0)",
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": -0,
          },
          Object {
            "scaleX": 0.0001,
          },
        ],
      }
    }
  />
</View>
`;

exports[`LinearProgress Component should render indeterminant without issues 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityValue={
    Object {
      "max": 1,
      "min": 0,
      "now": 0,
    }
  }
  accessible={true}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "blue",
        "height": 4,
        "overflow": "hidden",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Object {
        "backgroundColor": "rgb(255, 0, 0)",
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": -0,
          },
          Object {
            "scaleX": 0.0001,
          },
        ],
      }
    }
  />
</View>
`;

exports[`LinearProgress Component should render without issues 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityValue={
    Object {
      "max": 1,
      "min": 0,
      "now": 0,
    }
  }
  accessible={true}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "rgba(202, 113, 235, 0.4)",
        "height": 4,
        "overflow": "hidden",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Object {
        "backgroundColor": "#ca71eb",
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": -0,
          },
          Object {
            "scaleX": 0.0001,
          },
        ],
      }
    }
  />
</View>
`;
